### ISSUE TITLE
TEE_SetOperationKey(), TEE_SetOperationKey2() with invalid key-lengths

### ISSUE THREAD

**Issue by Hussain1811** [OP-TEE/optee_os#5040]
I was testing TEE_SetOperationKey2() for AES-XTS and TEE_SetOperationKey() for AES, DES, DES3 with an invalid key-length of 48bits. 
But the set key API returns TEE_SUCCESS.

I guess none of these algorithms accept a key-length of 48bits. So, shouldn't the set key api return TEE_ERROR_TARGET_DEAD?
(spec mentions: _Panic if the type, size, or usage of key is not compatible with the algorithm, mode, or size of the
operation_)

Environment used: OPTEE 3.15 over QEMU V8

**Comment by Hussain1811** [comment]
Could anyone help me with this doubt?

**Comment by jforissier** [comment]
I agree.
How did you create the key? Did you use `TEE_PopulateTransientObject()`? I think this function should probably reject invalid key lengths, too, although the specification is less explicit (Panic reason *"If the implementation detects any other error associated with this function that is not explicitly associated with a defined return code for this function"*).

**Comment by Hussain1811** [comment]
yes - i used TEE_PopulateTransientObject()
but it doesn't reject invalid keys and just returns TEE_SUCCESS

**Comment by jforissier** [comment]
OK. We could do this:
```diff
diff --git a/core/tee/tee_svc_cryp.c b/core/tee/tee_svc_cryp.c
index 44872415..b25e77c2 100644
--- a/core/tee/tee_svc_cryp.c
+++ b/core/tee/tee_svc_cryp.c
@@ -1609,6 +1609,9 @@ static TEE_Result tee_svc_cryp_obj_populate_type(
 			}
 			if (obj_size > o->info.maxKeySize)
 				return TEE_ERROR_BAD_STATE;
+			res = check_key_size(type_props, obj_size);
+			if (res != TEE_SUCCESS)
+				return TEE_ERROR_BAD_PARAMETERS;
 		}
 
 		/*
```
With this patch, `TEE_PopulateTransientObject()` would reject the invalid key size. I think it complies with the specification: *"The implementation can attempt to detect whether the attribute values are consistent; [...] If it detects invalid attributes, it SHALL return the error code TEE_ERROR_BAD_PARAMETERS and SHALL NOT panic."

**Comment by Hussain1811** [comment]
Makes sense.
i guess you could raise a PR with this patch.

**Comment by jforissier** [comment]
Unfortunately the patch makes quite a few test fail: regression_4006.39, regression_4007_dsa.1 and a bunch of tests from the GP suite. It seems all of them involve DSA but I haven't investigated further.

**Comment by Hussain1811** [comment]
Alright - no problem

**Comment by Hussain1811** [comment]
@jforissier should this issue be tagged as a bug?

**Comment by manishkanal1** [comment]
@jforissier I am working with @Hussain1811. we need some resolution on this issue. Should this be marked as a bug? or the are we expecting the API to return a success with an invalid key length, which does not make sense.

**Comment by jforissier** [comment]
@manishkanal1 I have marked this as a bug since this at least requires further investigation. This way the issue won't be marked "stale" nor closed automatically.
Would you have some time to figure out why the simple fix I proposed ([above](https://github.com/OP-TEE/optee_os/issues/5040#issuecomment-992447298)) causes errors in `xtest`?

**Comment by Hussain1811** [comment]
Sure - we shall look into this and revert back soon

**Comment by Hussain1811** [comment]
For the 4006.39 failure, i've seen that the algorithm identifier is 0x70004131 which is TEE_ALG_DSA_SHA256. But, on printing the key-size in check_key_size() function, it shows as 224. Shouldn't the key-size be 256 for TEE_ALG_DSA_SHA256 as i read somewhere that the key-size is tightly bound to the SHA? I guess that is why the test is failing.

**Comment by Hussain1811** [comment]
For the 4006.39 failure, i've seen that the algorithm identifier is 0x70004131 which is TEE_ALG_DSA_SHA256. But, on printing the key-size in check_key_size() function, it shows as 224. Shouldn't the key-size be 256 for TEE_ALG_DSA_SHA256 as i read somewhere that the key-size is tightly bound to the SHA? I guess that is why the test is failing.

**Comment by manishkanal1** [comment]
@jforissier @Hussain1811 has responded with the debugging results. Do we have a path forward here?

**Comment by jforissier** [comment]
> @jforissier **1)**For the 4006.39 failure, i've seen that the algorithm identifier is 0x70004131 which is TEE_ALG_DSA_SHA256. But, on printing the key-size in check_key_size() function, it shows as 224.

This value of 224 doesn't look correct indeed.
 
> **2)**For the 4007_dsa.1 failure, it is basically trying to generate a 1024bit DSA key. At the start of syscall_obj_generate_key() function --> key-size is 1024 just before calling tee_svc_obj_generate_key_dsa() function --> key-size is 1024 But, after tee_svc_obj_generate_key_dsa() --> tee_svc_cryp_obj_populate_type() --> check_key_size(): key-size is printed as 160 bits So, the key-size got modified from 1024 to 160 bits (unexpected?!?!) due to which we hit the same if block as (1) in check_key_size() function and hence a test-failure.

Here also, 160 looks wrong.

Please stay tuned, I think I found the bug and I'm testing a patch.

**Comment by jforissier** [comment]
> Please stay tuned, I think I found the bug and I'm testing a patch.

https://github.com/OP-TEE/optee_os/pull/5134

**Comment by jforissier** [comment]
https://github.com/OP-TEE/optee_os/pull/5135

**Comment by Hussain1811** [comment]
Tested with the bug-fixes provided in #5134 and #5135.
Results are as expected.

Hence, closing this issue as resolved.
